modules/dnsdist/collect.go

```diff
@@ -0,0 +1,77 @@
+package dnsdist
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/netdata/go.d.plugin/pkg/stm"
+	"github.com/netdata/go.d.plugin/pkg/web"
+)
+
+const (
+	urlPathLocalStatistics = "/jsonstat?command=stats"
+)
+
+func (d *DNSdist) collect() (map[string]int64, error) {
+	statistics, err := d.scrapeStatistics()
+	if err != nil {
+		return nil, err;
+	}
+
+	collected := make(map[string]int64)
+
+	d.collectStatistic(collected, statistics)
+	return collected, nil
+}
+
+func (d *DNSdist) collectStatistic(collected map[string]int64, statistics *statisticMetrics) {
+	for metric, value := range stm.ToMap(statistics) {
+		collected[metric] = int64(value)
+	}
+}
+
+func (d *DNSdist) scrapeStatistics() (*statisticMetrics, error) {
+	req, err := web.NewHTTPRequest(d.Request)
+	if err != nil {
+		return nil, err
+	}
+	
+	URL :=  []rune(d.Config.HTTP.Request.URL)
+	req.URL.Opaque = string(URL[5:]) + urlPathLocalStatistics
```

Comment:
```suggestion
    req.URL.Path = urlPathLocalStatistics
```

Reply:
Unfortunately, we cannot do this, because the `?` is converted to `%3F` and `dnsdist` webserver does not recognize the request delivering a `401` error.

Reply:
then see

```go
const urlPathJSONStat = "/jsonstat"

var urlQueryParamJSONStat = url.Values{"command": []string{"stats"}}.Encode()

func main() {
	req, _ := http.NewRequest("GET", "http://127.0.0.1:8081", nil)
	req.URL.Path = urlPathJSONStat
	req.URL.RawQuery = urlQueryParamJSONStat
	fmt.Println(req.URL)
}
```

> http://127.0.0.1:8081/jsonstat?command=stats

Reply:
The same error:

```
2020-11-23 16:49:16: goplugin ERROR: base[base] 'http://127.0.0.1:8083/jsonstat?command=stats' returned HTTP status code: 401
2020-11-23 16:49:16: goplugin ERROR: base[base] 'http://127.0.0.1:8083/jsonstat?command=stats' returned HTTP status code: 401
--- FAIL: Test_Check (0.00s)
    --- FAIL: Test_Check/success (0.00s)
        dnsdist_test.go:93: 
            	Error Trace:	dnsdist_test.go:93
            	Error:      	Should be true
            	Test:       	Test_Check/success
FAIL
FAIL	github.com/netdata/go.d.plugin/modules/dnsdist	0.003s
FAIL
```

Reply:
401 is `Unauthorized`, it should has nothing to do with path, i will check it myself.

Reply:
I agree, but with the current code it is working fine.

Reply:
current code is what make me trying to imrpove it, i dont think `:5` is good

Reply:
I will take a look in the package to bring a better solution, but a curiosity, what is the problem with string concatenation at a variable that is not encoded?

Reply:
It is better use query params if you want to set query params and not hacky way (go plugin would panic if i set `url` with len(something) < 5 ).

---

ok, i installed `v1.5.1` in a docker container, it works

```go
	req, _ := http.NewRequest("GET", "http://127.0.0.1:8083", nil)
	req.SetBasicAuth("#", "pass")

	req.URL.Path = "/jsonstat"
	req.URL.RawQuery = url.Values{"command": []string{"stats"}}.Encode()
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()

	var m map[string]interface{}

	if err := json.NewDecoder(resp.Body).Decode(&m); err != nil {
		log.Println(err)
		return
	}
	bs, _ := json.MarshalIndent(m, "", " ")

	fmt.Println(resp.StatusCode)
	
	fmt.Println(string(bs))
```

Response:

```
200
{
 "acl-drops": 0,
 "cache-hits": 0,
 "cache-misses": 0,
 "cpu-iowait": 839,
 "cpu-steal": 0,
 "cpu-sys-msec": 843,
 "cpu-user-msec": 2531,
 "doh-query-pipe-full": 0,
 "doh-response-pipe-full": 0,
 "downstream-send-errors": 0,
 "downstream-timeouts": 0,
 "dyn-block-nmg-size": 0,
 "dyn-blocked": 0,
 "empty-queries": 0,
 "fd-usage": 19,
 "frontend-noerror": 0,
 "frontend-nxdomain": 0,
 "frontend-servfail": 0,
 "latency-avg100": 0,
 "latency-avg1000": 0,
 "latency-avg10000": 0,
 "latency-avg1000000": 0,
 "latency-count": 0,
 "latency-slow": 0,
 "latency-sum": 0,
 "latency0-1": 0,
 "latency1-10": 0,
 "latency10-50": 0,
 "latency100-1000": 0,
 "latency50-100": 0,
 "no-policy": 0,
 "noncompliant-queries": 0,
 "noncompliant-responses": 0,
 "over-capacity-drops": 0,
 "packetcache-hits": 0,
 "packetcache-misses": 0,
 "queries": 0,
 "rdqueries": 0,
 "real-memory-usage": 44265472,
 "responses": 0,
 "rule-drop": 0,
 "rule-nxdomain": 0,
 "rule-refused": 0,
 "rule-servfail": 0,
 "security-status": 1,
 "self-answered": 0,
 "server-policy": "leastOutstanding",
 "servfail-responses": 0,
 "too-old-drops": 0,
 "trunc-failures": 0,
 "udp-in-errors": 0,
 "udp-noport-errors": 0,
 "udp-recvbuf-errors": 0,
 "udp-sndbuf-errors": 0,
 "uptime": 859
}
```

Reply:
I changed the algorithm now and it works.